network/src/protocols/discovery/mod.rs

```diff
@@ -408,54 +452,81 @@ async fn handle_inbound_substream<TSubstream>(
-                .map(|note| {
-                    is_valid(&note.to_owned(), trusted_peers.clone()).map_err(|e| {
-                        security_log(SecurityEvent::InvalidNetworkPeer)
-                            .error(&e)
-                            .data(&note)
-                            .data(&trusted_peers)
-                            .log();
-                        e
-                    })
-                })
-                .collect::<Result<Vec<_>, NetworkError>>()?;
-            Ok(msg.take_notes().into_vec())
-        });
-    (peer_id, result)
+        .await;
+
+    let res_notes = res_msg.and_then(|mut msg| {
+        msg.get_notes().iter().try_for_each(|note| {
+            is_valid(&note, &trusted_peers).map_err(|err| {
+                security_log(SecurityEvent::InvalidNetworkPeer)
+                    .error(&err)
+                    .data(&peer_id)
+                    .data(&note)
+                    .data(&trusted_peers)
+                    .log();
+                err
+            })
+        })?;
+        Ok(msg.take_notes().into_vec())
+    });
+
+    (peer_id, res_notes)
 }
 
 // Verifies validity of notes. Following conditions should be met for validity:
 // 1. We should be able to correctly parse the peer id in each note.
-// 2. The signature should be verified to be of the given peer for the serialized peer info.
+// 2. The signature of the serialized peer info should be valid for the given peer_id.
 // 3. The address(es) in the PeerInfo should be correctly parsable as Multiaddrs.
+// 4. The signature of the serialized full node payload should be valid for the given peer_id.
 fn is_valid(
     note: &Note,
-    trusted_peers: Arc<RwLock<HashMap<PeerId, NetworkPublicKeys>>>,
+    trusted_peers: &RwLock<HashMap<PeerId, NetworkPublicKeys>>,
 ) -> Result<(), NetworkError> {
+    // validate PeerId
+
     let peer_id = PeerId::try_from(note.get_peer_id())
         .map_err(|err| err.context(NetworkErrorKind::ParsingError))?;
-    verify_signatures(
-        trusted_peers,
-        peer_id,
-        note.get_signature(),
-        note.get_peer_info(),
-    )?;
-    let peer_info: PeerInfo = protobuf::parse_from_bytes(note.get_peer_info())?;
+
+    // validate PeerInfo
+
+    if !note.has_signed_peer_info() {
+        return Err(format_err!("Discovery Note missing signed_peer_info field")
+            .context(NetworkErrorKind::ParsingError)
+            .into());
+    }
+    let signed_peer_info = note.get_signed_peer_info();
+
+    let peer_info_bytes = signed_peer_info.get_peer_info();
+    let peer_info_signature = signed_peer_info.get_signature();
+    verify_signature(trusted_peers, peer_id, peer_info_signature, peer_info_bytes)?;
+
+    let peer_info: PeerInfo = protobuf::parse_from_bytes(peer_info_bytes)?;
     for addr in peer_info.get_addrs() {
         let _: Multiaddr = Multiaddr::try_from(addr.clone())?;
     }
+
+    // validate FullNodePayload
+
+    if !note.has_signed_full_node_payload() {
```

Comment:
I'm wondering if this should be an error condition - what if the node does not have any payload to share or does not want to forward the payload of other peers?